### PR TITLE
feat: mark programmed=false with an error when too many addresses are assigned

### LINKED ISSUE #3662: Setting EnvoyService to NodePort type has a bug when total number of k8s nodes > 16
*Description*:
When `EnvoyService` is set to type `NodePort` we are seeing the following error:
> status.addresses: Too many. Must have at most 16 items

When the service is of type NodePort, the address is set to all of the node addresses, as seen here:
https://github.com/envoyproxy/gateway/blob/c752b889e81190d8c4dd81eafb71dfa5e5110311/internal/gatewayapi/status/gateway.go#L73

*Repro steps*:
1. Have a Kubernetes environment with > 16 nodes
2. Set EnvoyService type to NodePort:
```
    envoyService:
      type: NodePort
```
3. Read logs from envoy gateway

*Environment*:
Gateway version: v1.0.1
Envoy version: v1.29

*Logs*:
```
2024-06-24T15:18:20.599Z	ERROR	provider	kubernetes/status_updater.go:115	unable to update status	{"runner": "provider", "name": "gateway", "namespace": "...", "error": "Gateway.gateway.networking.k8s.io "gateway" is invalid: status.addresses: Too many: 67: must have at most 16 items"}
```

### PR DESCRIPTION
**Pull request by sanposhiho**
**What type of PR is this?**

**What this PR does / why we need it**:

This PR changes the impl to mark programmed=false with an error when too many addresses are assigned.
The motivation here is to enhance the visibility of this error to users.

**Which issue(s) this PR fixes**:

Fixes #3662

### CODE REVIEW

File: internal/gatewayapi/status/gateway.go [hunk 1]
```diff
@@ -137,6 +137,13 @@ func computeGatewayProgrammedCondition(gw *gwapiv1.Gateway, deployment *appsv1.D
 			"No addresses have been assigned to the Gateway", time.Now(), gw.Generation)
 	}
 
+	if len(gw.Status.Addresses) > 16 {
+		return newCondition(string(gwapiv1.GatewayConditionProgrammed), metav1.ConditionFalse,
```
Comment (arkodg) [inline internal/gatewayapi/status/gateway.go:141]: thanks for improving the status message !
we should also discard `gw.Status.Addresses[16:len(gw.Status.Addresses)-1]` so the status update doesnt fail when writing into k8s API server

File: internal/gatewayapi/status/gateway.go [hunk 2]
```diff
@@ -96,8 +96,15 @@ func UpdateGatewayStatusProgrammedCondition(gw *gwapiv1.Gateway, svc *corev1.Ser
 	} else {
 		gw.Status.Addresses = nil
 	}
+
 	// Update the programmed condition.
 	gw.Status.Conditions = MergeConditions(gw.Status.Conditions, computeGatewayProgrammedCondition(gw, deployment))
+
+	if len(gw.Status.Addresses) > 16 {
```
Comment (arkodg) [inline internal/gatewayapi/status/gateway.go:103]: can we move the truncation and the status message in one place to easier to reason ?

**Review by sanposhiho — COMMENTED** [review]
followup comments

File: internal/gatewayapi/status/gateway.go [hunk 3]
```diff
@@ -128,26 +129,45 @@ func computeGatewayAcceptedCondition(gw *gwapiv1.Gateway, accepted bool) metav1.
...
+	messageAddressNotAssigned  = "No addresses have been assigned to the Gateway"
+	messageFmtTooManyAddresses = "Too many addresses (%d) have been assigned to the Gateway, the maximum number of addresses is 16"
+	messageNoResources         = "Deployment replicas unavailable"
+	messageFmtProgrammed       = "Address assigned to the Gateway, %d/%d envoy Deployment replicas available"
+)
+
 // computeGatewayProgrammedCondition computes the Gateway Programmed status condition.
 // Programmed condition surfaces true when the Envoy Deployment status is ready.
-func computeGatewayProgrammedCondition(gw *gwapiv1.Gateway, deployment *appsv1.Deployment) metav1.Condition {
+func computeGatewayProgrammedCondition(gw *gwapiv1.Gateway, deployment *appsv1.Deployment) {
```
[the remark this answers is not in the document]
Comment (arkodg) [reply internal/gatewayapi/status/gateway.go:141]: if that case lets call this `updateGatewayProgrammedCondition` ?

[the remark this answers is not in the document]
Comment (sanposhiho) [reply internal/gatewayapi/status/gateway.go:141]: That sounds better. Updated. 🙏

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Review by shawnh2 — APPROVED** [review]
lgtm

**Pull request merged**